test_conformance/extensions/cl_khr_command_buffer/command_buffer_queue_substitution.cpp

```diff
@@ -0,0 +1,321 @@
+            ADD_PROP(CL_QUEUE_PROFILING_ENABLE)
+        };
+
+        for (auto&& prop : props)
+        {
+            check_property(prop);
+            test_error_ret(error, "clCreateCommandQueue failed", ret_queue);
+            if (ret_queue != nullptr) return ret_queue;
+        }
+
+        return ret_queue;
+    }
+
+    //--------------------------------------------------------------------------
+    cl_int SetUp(int elements) override
+    {
+        // By default command queue is created without properties,
+        // if test requires queue with properties default queue must be
+        // replaced.
+        if (properties_use_requested)
+        {
+            // due to the skip condition
+            cl_int error = CL_SUCCESS;
+            queue = CreateCommandQueueWithProperties(error);
+            test_error(error, "CreateCommandQueueWithProperties failed");
+
+            cl_command_queue_properties cqp;
+            error = clGetCommandQueueInfo(queue, CL_QUEUE_PROPERTIES,
+                                          sizeof(cqp), &cqp, NULL);
+            test_error(error, "clGetCommandQueueInfo failed");
+
+            if (simultaneous_use_support
+                && (cqp & CL_QUEUE_OUT_OF_ORDER_EXEC_MODE_ENABLE))
+            {
+                log_info(
+                    "Queue property CL_QUEUE_OUT_OF_ORDER_EXEC_MODE_ENABLE "
+                    "not supported with simultaneous use in this test\n");
+                return CL_INVALID_QUEUE_PROPERTIES;
+            }
+        }
+
+        return BasicCommandBufferTest::SetUp(elements);
+    }
+
+    //--------------------------------------------------------------------------
+    cl_int Run() override
+    {
+        // record command buffer with primary queue
+        cl_int error = RecordCommandBuffer();
+        test_error(error, "RecordCommandBuffer failed");
+
+        // create substitute queue
+        clCommandQueueWrapper new_queue;
+        if (properties_use_requested)
+        {
+            new_queue = CreateCommandQueueWithProperties(error);
+            test_error(error, "CreateCommandQueueWithProperties failed");
+        }
+        else
+        {
+            const cl_command_queue_properties queue_properties = 0;
+            new_queue =
+                clCreateCommandQueue(context, device, queue_properties, &error);
+            test_error(error, "clCreateCommandQueue failed");
+        }
+
+        if (simultaneous_use_support)
+        {
+            // enque simultaneous command-buffers with substitute queue
+            error = RunSimultaneous(new_queue);
+            test_error(error, "RunSimultaneous failed");
+        }
+        else
+        {
+            // enque single command-buffer with substitute queue
+            error = RunSingle(new_queue);
+            test_error(error, "RunSingle failed");
+        }
+
+        if (properties_use_requested)
+        {
+            clReleaseCommandQueue(queue);
```

Comment:
I'm not quite following why the `queue` needs to be explicitly released here.  Is this needed?  If so, can we use a queue wrapper for it instead so an explicit release is not required?

Reply:
I think this is the queue coming from the main test entry-point, which is then stored in `BasicCommandBufferTest` as a `cl_command_queue`. We don't use a wrapper if I recall because the common test fixture code which created that original queue will do the release after the test returns. 

This `BasicCommandBufferTest::queue` member variable is then overwritten in `SubstituteQueueTest::SetUp` before calling `BasicCommandBufferTest::SetUp` to create a command-buffer from a queue with properties set. So here we're releasing that new queue we created in `SubstituteQueueTest::SetUp` 

I agree this is confusing, and the way that comes to mind for improving this is by creating a `SubstituteQueueTest` command-buffer member variable in `SubstituteQueueTest::SetUp` which we can use in the tests, rather than using the base class command-buffer variable. We do this already in `OutOfOrderTest ` with the `out_of_order_command_buffer` variable, so can use that for reference.

Reply:
OK, yes, I get it now.

Could we do this though?

1. Use a clCommandQueueWrapper instead of a cl_command_queue for BasicCommandBufferTest::queue.
2. Explicitly "retain" the command queue in the BasicCommandBufferTest constructor.  This is needed because we're using the command queue from the harness and we don't necessarily want the command queue to be destroyed when the wrapper is destroyed, we just want to release it (decrement the refcount by one).
3. Now that we're using a command queue wrapper we can remove this explicit release.

I just gave this a try and it seemed to work for both this "properties" test and the other command buffer tests that do not replace the queue, with no leaks reported.

Reply:
That's a good idea, using an explicit `clRetainCommandQueue` and then holding the queue in a wrapper I agree would be a better solution.

Reply:
Done